airflow/jobs.py

```diff
@@ -403,39 +422,67 @@ def schedule_dag(self, dag):
+                    )
+            ).order_by(DagRun.execution_date.desc()).first()
+
+            if previous:
+                self.logger.debug("Previous run {}".format(previous.execution_date))
+
+            # Determine next run date
             next_run_date = None
-            if dag.schedule_interval == '@once' and not last_scheduled_run:
+            if dag.schedule_interval == '@once' and not previous:
                 next_run_date = datetime.now()
-            elif not last_scheduled_run:
+            elif not previous:
                 # First run
-                TI = models.TaskInstance
-                latest_run = (
-                    session.query(func.max(TI.execution_date))
-                    .filter_by(dag_id=dag.dag_id)
-                    .scalar()
-                )
-                if latest_run:
-                    # Migrating from previous version
-                    # make the past 5 runs active
-                    next_run_date = dag.date_range(latest_run, -5)[0]
+                self.logger.debug("Scheduling {} for the first time".format(dag.dag_id))
+
+                # Set start_date based on tasks
+                task_start_dates = [t.start_date for t in dag.tasks]
+                if task_start_dates:
+                    # set next_run_date to start_date + interval unless start date
+                    # is on the interval
+                    min_task_start_date = min(task_start_dates)
+                    next_run_date = dag.following_schedule(min_task_start_date)
+                    next_run_date = self._normalize_schedule(dag,
+                                                             next_run_date,
+                                                             min_task_start_date)
+                    self.logger.debug("Next run date based on tasks {}"
+                                      .format(next_run_date))
                 else:
-                    task_start_dates = [t.start_date for t in dag.tasks]
-                    if task_start_dates:
-                        next_run_date = min(task_start_dates)
-                    else:
-                        next_run_date = None
+                    next_run_date = None
             elif dag.schedule_interval != '@once':
-                next_run_date = dag.following_schedule(last_scheduled_run)
+                next_run_date = dag.following_schedule(previous.execution_date)
 
             # don't ever schedule prior to the dag's start_date
+            # this can happen if the tasks have a start date that
+            # is before the dag start date
             if dag.start_date:
                 next_run_date = dag.start_date if not next_run_date else max(next_run_date, dag.start_date)
+                if next_run_date == dag.start_date:
+                    next_run_date = dag.following_schedule(dag.start_date)
+                    next_run_date = self._normalize_schedule(dag,
```

Comment:
With `DAG.normalize_schedule` I suggested above you'd have a single simpler line of code here instead of 2 lines